crates/rust-analyzer/src/to_proto.rs

```diff
@@ -656,22 +656,14 @@ pub(crate) fn resolved_code_action(
 pub(crate) fn runnable(
     snap: &GlobalStateSnapshot,
     file_id: FileId,
-    runnable: Runnable,
+    runnable: &Runnable,
```

Comment:
I *think* semantically this should be `runnable: Runnable`. It's nice to statically enforce that we don't duplicate runnables. The lifetimes work our if you make `HoverResult::actions` public, which seems like a good idea -- so far it seems like ide-layer API work best if they are just plain data. 

It might make sense to make `results` public as well, but I'd say that they shold be wholesale replaced with something more specific (though this obviously should be handled by an unrelated PR):

```rust
struct HoverResult {
  ty: Option<Stirng>,
  docs: Option<String>,
  module: Option<String>,
}
```

Reply:
Might be on this level (ra_ide::HoverResult) we can even have not only a string, but something more specific (Type, Function, Local, etc.)?

Reply:
Heh, looks like the strategy of documenting knowledge works :)

https://github.com/rust-analyzer/rust-analyzer/pull/4792/commits/ee8dec5dc11cfecf219b6510b0eadd9691a82ba5

TL;DR: eventually, we *have* to show the info in the editor, where compiler concepts like definition and usages do not work (but, eg, strings and icons do). The `ra_ide` is exactly the layer where we switch from "compiler view" to the "editor" view, so in general ide speaks in generalized `Display` things.

Reply:
it does :)

Got it. Thanks.